larq/optimizers_v2.py

```diff
@@ -5,6 +5,84 @@
 from copy import deepcopy
 
 
+@utils.register_keras_custom_object
+class OptimizerGroup(tf.keras.optimizers.Optimizer):
```

Comment:
An alternative to making the implementation general is to make it clear this is a specific case, e.g. by naming the class "BNNOptimizerDuo". In this approach I would approach largely as is, on the condition that the decision rule for whether a variable is binary can be made more robust.

If the decision rule isn't robust, I think it's better if the user gets a responsibility to set the rule. Currently, passing a name to a quantized layer introduces a rather subtle bug.

(I still have a slight preference for an abstract `OptimizerGroup`, possibly with a specific `BNNOptimizerDuo` as subclass, which would trivially extend to other/more bit widths and not add much complexity in my view. But this approach would also be fine I think.)

Reply:
I've renamed `OptimizerGroup` to `BNNOptimizerDuo` in f65d7a1. I think we should wait to add a more general superclass around this until we have more subclasses (like a Trio with a ternary optimizer).

I agree about the subtle bugs that hiding the `_is_binary` rule in this class could introduce for users not following our binary layer name conventions; we can check back when we know whether the `tf.Variable` subclass decision rule is viable.